### PR TITLE
fix bugs in circular reference detection

### PR DESCRIPTION
**Pull request by bourdakos1**
There were a few cases with the old cycle detection algorithm that would cause it to infinitely recurse. This PR instead uses a loop to make it easier to timeout on long checks or an unforeseen infinite loop. I also added some tests to check for some common cycle patterns and tests for where the previous algorithm was failing. 
 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

**Comment by lresende** [comment]
@bourdakos1 while you are at it, does it make sense to start decoupling the validation code from the widget and start thinking how it would change its behavior when different runtimes are selected (see #870). And these are definitely different/new prs.

File: packages/pipeline-editor/src/validation.spec.ts [hunk 1]
```diff
@@ -0,0 +1,343 @@
+/*
```
Comment (lresende) [inline packages/pipeline-editor/src/validation.spec.ts:1]: For consistency, should tests be placed under the test directory?

**Comment by bourdakos1** [comment]
@lresende I've already decoupled validation from the widget in my large PR so that will come eventually. The behavior in the large PR is that it always performs cycle checks no matter the runtime, but then all other checks come from the node schema. So for the most part the validation is decided by the node, not the runtime. However, the runtime plugin decides what nodes to serve and could modify the schema to have different checks.

I think "running locally" is a special case and could be part of the node schema. for example instead of:
```
- id: runtime_image
  type: string
  required: true
  ...
```

We could have something like:
```
- id: runtime_image
  type: string
  required:
    runtime: true
    local: false
  ...
```

File: packages/pipeline-editor/src/PipelineEditorWidget.tsx [hunk 2]
```diff
@@ -1125,32 +1104,30 @@ export class PipelineEditor extends React.Component<
    * @returns null if pipeline is valid, error message if not.
    */
   validateAllLinks(): string {
```
Comment (lresende) [inline packages/pipeline-editor/src/PipelineEditorWidget.tsx:1106]: Thank you for starting this cleanup, at some point I really would like to see one call in the widget, which is PipelineService.validatePipeline(....) and everything else is driven from there (with calling other utility classes).

File: packages/pipeline-editor/src/validation.ts [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// 3 seconds should be ample time.
+// should only protect against any infinite loops.
+const TIMEOUT = 3 * 1000;
```
Comment (lresende) [inline packages/pipeline-editor/src/validation.ts:19]: When do you expect to get here? If we are trying to identify infinite loops, should we stop when we find the first occurrence of trying to visit a vertice that has been already processed as that identifies a circular dependency?

File: packages/pipeline-editor/src/validation.ts [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+export interface ILink {
+  id: string;
+  trgNodeId: string;
+  srcNodeId: string;
+  type: string;
+}
+
+export const checkCircularReferences = (links: ILink[]): string[] => {
```
Comment (lresende) [inline packages/pipeline-editor/src/validation.ts:28]: Could you add a quick "function doc" on what it does?

File: packages/pipeline-editor/src/validation.spec.ts [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+  // │ a │──l0─►│ b │──l1─►│ c │──l5─►│ f │──l6─►│ g │
+  // ╰───╯      ╰───╯      ╰───╯      ╰───╯      ╰───╯
+  //                         │          ▲
+  //                        l2          l4
+  //                         ▼          │
+  //                       ╭───╮      ╭───╮
+  //                       │ d │──l3─►│ e │
+  //                       ╰───╯      ╰───╯
+  {
+    it: 'should not have a bug that a previous implementation had',
```
Comment (lresende) [inline packages/pipeline-editor/src/validation.spec.ts:290]: what was the bug? That might be a better description for someone seeing this for the first time...

File: packages/pipeline-editor/src/validation.spec.ts [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+  //                         │
+  //                        l4
+  //                         ▼
+  //                       ╭───╮      ╭───╮      ╭───╮      ╭───╮      ╭───╮
+  //                       │ f │──l5─►│ g │──l6─►│ h │──l7─►│ i │──l8─►│ j │
+  //                       ╰───╯      ╰───╯      ╰───╯      ╰───╯      ╰───╯
+  //                                                          ▲          │
+  //                                                          ╰────l9────╯
+  {
+    it: 'should handle long forks',
```
Comment (lresende) [inline packages/pipeline-editor/src/validation.spec.ts:236]: for the algorithm, what is the difference between this scenario and the one described on the test below?

File: packages/pipeline-editor/src/PipelineEditorWidget.tsx (hunk 2 above)
Comment (bourdakos1) [reply packages/pipeline-editor/src/PipelineEditorWidget.tsx:1106]: 👍 yea that's how it is in the large PR

File: packages/pipeline-editor/src/validation.ts (hunk 3 above)
Comment (bourdakos1) [reply packages/pipeline-editor/src/validation.ts:19]: There should never never be an infinite loop (unless there is a bug my the algorithm) this is just a safety net for that case

File: packages/pipeline-editor/src/validation.spec.ts (hunk 6 above)
Comment (bourdakos1) [reply packages/pipeline-editor/src/validation.spec.ts:236]: When the algorithm gets to node `C` it either traverses down `l2` or `l4` first depending on the order of the links, this just checks that it doesn't matter which it traverses first

File: packages/pipeline-editor/src/validation.spec.ts (hunk 5 above)
Comment (bourdakos1) [reply packages/pipeline-editor/src/validation.spec.ts:290]: I actually don't know haha... I just had this graph sketched out on a piece of paper of other graphs that had issues and couldn't remember what the issue was when writing tests so I gave it a vague name 😬

It either incorrectly indicated a cycle or had caused an infinite loop, but also could have not been and bug and wrote it down for another reason 🤷‍♂️

**Comment by marthacryan** [comment]
Good catch! +1 on all of Luciano's comments (especially placement of spec file) but otherwise this is a good addition.

**Review by marthacryan — APPROVED** [review]

**Review by lresende — APPROVED** [review]

**Pull request merged**